core/chains/evm/logpoller/orm.go

```diff
@@ -16,23 +16,65 @@ import (
+	SelectLatestBlock(qopts ...pg.QOpt) (*LogPollerBlock, error)
+
+	SelectLogs(start, end int64, address common.Address, eventSig common.Hash, qopts ...pg.QOpt) ([]Log, error)
+	SelectLogsWithSigs(start, end int64, address common.Address, eventSigs []common.Hash, qopts ...pg.QOpt) ([]Log, error)
+	SelectLogsCreatedAfter(eventSig common.Hash, address common.Address, after time.Time, confs int, qopts ...pg.QOpt) ([]Log, error)
+	SelectLatestLogByEventSigWithConfs(eventSig common.Hash, address common.Address, confs int, qopts ...pg.QOpt) (*Log, error)
+	SelectLatestLogEventSigsAddrsWithConfs(fromBlock int64, addresses []common.Address, eventSigs []common.Hash, confs int, qopts ...pg.QOpt) ([]Log, error)
+	SelectLatestBlockByEventSigsAddrsWithConfs(fromBlock int64, eventSigs []common.Hash, addresses []common.Address, confs int, qopts ...pg.QOpt) (int64, error)
+
+	SelectIndexedLogs(address common.Address, eventSig common.Hash, topicIndex int, topicValues []common.Hash, confs int, qopts ...pg.QOpt) ([]Log, error)
+	SelectIndexedLogsByBlockRange(start, end int64, address common.Address, eventSig common.Hash, topicIndex int, topicValues []common.Hash, qopts ...pg.QOpt) ([]Log, error)
+	SelectIndexedLogsCreatedAfter(address common.Address, eventSig common.Hash, topicIndex int, topicValues []common.Hash, after time.Time, confs int, qopts ...pg.QOpt) ([]Log, error)
+	SelectIndexedLogsTopicGreaterThan(address common.Address, eventSig common.Hash, topicIndex int, topicValueMin common.Hash, confs int, qopts ...pg.QOpt) ([]Log, error)
+	SelectIndexedLogsTopicRange(address common.Address, eventSig common.Hash, topicIndex int, topicValueMin, topicValueMax common.Hash, confs int, qopts ...pg.QOpt) ([]Log, error)
+	SelectIndexedLogsWithSigsExcluding(sigA, sigB common.Hash, topicIndex int, address common.Address, startBlock, endBlock int64, confs int, qopts ...pg.QOpt) ([]Log, error)
+	SelectLogsDataWordRange(address common.Address, eventSig common.Hash, wordIndex int, wordValueMin, wordValueMax common.Hash, confs int, qopts ...pg.QOpt) ([]Log, error)
+	SelectLogsDataWordGreaterThan(address common.Address, eventSig common.Hash, wordIndex int, wordValueMin common.Hash, confs int, qopts ...pg.QOpt) ([]Log, error)
+	SelectLogsUntilBlockHashDataWordGreaterThan(address common.Address, eventSig common.Hash, wordIndex int, wordValueMin common.Hash, untilBlockHash common.Hash, qopts ...pg.QOpt) ([]Log, error)
+}
+
+type DbORM struct {
 	chainID *big.Int
 	q       pg.Q
 }
 
-// NewORM creates an ORM scoped to chainID.
-func NewORM(chainID *big.Int, db *sqlx.DB, lggr logger.Logger, cfg pg.QConfig) *ORM {
+// NewORM creates an DbORM scoped to chainID.
+func NewORM(chainID *big.Int, db *sqlx.DB, lggr logger.Logger, cfg pg.QConfig) *DbORM {
 	namedLogger := lggr.Named("Configs")
 	q := pg.NewQ(db, namedLogger, cfg)
-	return &ORM{
+	return &DbORM{
 		chainID: chainID,
 		q:       q,
 	}
 }
 
+func (o *DbORM) Q() pg.Q {
+	return o.q
```

Comment:
Unfortunately, it's not that easy at this moment. LogPoller is still tightly coupled to the underlying Postgres. I know it's a leaky abstraction here, but it's good enough now. The purpose of this PR is not to decouple the persistence layer from the business logic in the LogPoller but to introduce better observability. I can make it more explicit in the comment for this function.

Reply:
IMO, supporting anything but postgres would be a bad direction to head in for multiple reasons.  I've been trying to make the case for a while that we should move in the opposite direction and become more opinionated about what database version and setup they have rather than moving to a more "anything goes" approach.

There were a large number of incidents and war rooms spawned from the decision to allow node ops to run postgres on AWS rather than by itself or in docker.  The reason for that is that it makes the environment very different from the one that we test in. If node ops were running mysql or Oracle or something else, I can't imagine how many issues we'd run into. There's just no way to optimize Chainlink for an arbitrary database... we want to keep it as specific as possible so the results when it runs are repeatable rather than unpredictable.

Reply:
Good point @reductionista. Agree that we should not let NOPs freely decide on such a critical component like LogPoller and the database engine. I can only imagine how many issues it can cause in a production environment, especially if we consider that we have limited capacity for testing it. What I'm trying to say is that layered architecture is better for us as developers and architects of this system. Having a loosely coupled and more abstracted solution in which ORM can implement any technology under the hood gives us more space for improvements in the future. This way, we can, at some point, decide that maybe key-value storage would be a better persistence layer for LogPoller or maybe we should keep this data in memory instead of persisting them for some high-throughput products that also require low latency. ORM interface, which doesn't expose implementation details like Postgres internals, should give us this kind of flexibility to replace the engine that is under the hood. I'm not saying we need it now, but probably with little effort, we can have that option for future generations and keep the solution more aligned with the open-close principle

Reply:
Long story short, if all Postgres-specific details and logic are hidden behind the ORM implementation, for instance we could switch to storing EVM logs in memory without any changes (or little changes) to the LogPoller. Just implement `InMemoryORM` and inject it into LogPoller ;)

Reply:
Oh, yes I agree that having an in-memory option would be a good motivation for having the ORM be more general.

There are a host of issues with that, mostly surrounding how we can guarantee not missing logs when the node crashes and comes back up later... but most of those we'll have to solve anyway once we switch to separate LOOPP plugins for core & relay.  (Currently, we use pg features to guarantee atomicity between job and filter creation/deletion and filter persistence to solve issues we saw with missing logs when filters were only stored in-memory.  After the LOOPP plugin separation happens, those tables will be in separate db's so there's no way to guarantee they stay in sync... we'll just have to have some kind of cleanup process if they get out of sync.)